standard/expressions.md

```diff
@@ -2227,17 +2228,103 @@ If the *primary_no_array_creation_expression* of an *element_access* is a value
 
 #### 12.8.12.2 Array access
 
-For an array access, the *primary_no_array_creation_expression* of the *element_access* shall be a value of an *array_type*. Furthermore, the *argument_list* of an array access is not allowed to contain named arguments. The number of expressions in the *argument_list* shall be the same as the rank of the *array_type*, and each expression shall be of type `int`, `uint`, `long`, or `ulong,` or shall be implicitly convertible to one or more of these types.
+For an array access, the *primary_no_array_creation_expression* of the *element_access* shall be a value of an *array_type*. Furthermore, the *argument_list* of an array access shall not contain named arguments. The number of expressions in the *argument_list* shall be the same as the rank of the *array_type*, and each expression shall be of type `int`, `uint`, `long`, or `ulong`, or shall be implicitly convertible to one or more of these types.
 
-The result of evaluating an array access is a variable of the element type of the array, namely the array element selected by the value(s) of the expression(s) in the *argument_list*.
+For single-dimension array access the argument expression can also be of type `System.Index` or `System.Range`, or implicitly convertible to one of these.
+
+See ([§12.8.17.3](expressions.md#128173-object-initializers) for a constraint on using `System.Index` in the context of an *initializer_target*.
+
+The result of evaluating an array access that does not involve `System.Range` is a variable of the element type of the array, namely the array element selected by the value(s) of the expression(s) in the *argument_list*.
+
+The result of evaluating an array access that involves `System.Range` is a slice (§indexable-sequence-general) of the array being accessed, as selected by the value(s) of the expression(s) in the *argument_list*. The resulting slice’s element type is the same as the element type of the array being accessed.
```

Comment:
We may need to differentiate between compile-time and runtime types here. Does the code below throw an exception?

```csharp
object[] obj = new string[5];
object[] slice = obj[0..5];
slice[0] = new object();
```

Reply:
(Spoiler alert: it throws an exception.)

Reply:
This concurs with Joseph's suggestion, for the removal of any doubt.

Reply:
And this example prints "System.String[]":

```
string[] source = ["one", "two", "three", "four"] ;

object[] basic = source;

var slice = basic[0..2];
Type t = slice.GetType();
Console.WriteLine(t);
```